### ISSUE TITLE
Development plan for v.0.4.0

### ISSUE THREAD

**Issue by sw005320** [espnet/espnet#650]
There are a lot of exciting on-going PRs.
I'm making an issue of summarizing development items for v.0.4.0.
Please let us know if you have any request for the major update.
If you guys are interested in helping the development, that is also great!

- [x] multichannel end-to-end #596 
- [x] SpecAugment #734 
- [ ] neural machine translation #563 
- [x] transformer #555 #465 #690 
- [x] transformer TTS #752 
- [ ] transformer LM
- [ ] RNN transducer 
- [ ] cycle consisntency based semi-supervised training #712 
- [ ] online decoding #638 #645 
- [ ] speed up for decoding #748 
- [ ] tutorial based on [Google colab](https://colab.research.google.com/notebooks/welcome.ipynb)
- [ ] asynchronous snapshot saving
- add more recipes (see https://github.com/espnet/espnet/wiki/How-to-port-the-Kaldi-recipe-to-the-ESPnet-recipe%3F when you port the Kaldi recipe to the ESPnet recipe)
    - [x] TIMIT #698 
    - [x] JNAS #725
    - [ ] icsi meeting (https://github.com/kaldi-asr/kaldi/pull/2435)
    - [ ] jsut ASR #793 
    - [x] tedlium v3 #739 
    - [ ] common voice
    - [x] aurora4 #770 
    - [x] Russian open STT recipe # 768 
    - [x] libri tts #795 
- update major recipes with SOTA for several ASR benchmarks
    - [x] tedlium #691 #803 
    - [x] Librispeech #740 
    - [ ] SWB
    - [ ] WSJ
    - [x] CSJ #782 #737 
    - [x] voxforge Italian (Adi set) #758 
- [x] only support python3 #691 
- [x] use pytorch 1.0.1 #721 
- [x] use chainer 6.0.0 #753 
- [x] support config files instead of too many options #722 
- [ ] completely fix pytorch warp_ctc
- [ ] Kaldi-free
- [ ] publish each version release with Zenodo (https://zenodo.org/)

**Comment by ShigekiKarita** [comment]
The End Of Life date (EOL, sunset date) for Python 2.7 is Jan. 1 2020 
https://legacy.python.org/dev/peps/pep-0373/
https://www.numpy.org/neps/nep-0014-dropping-python2.7-proposal.html

**Comment by ShigekiKarita** [comment]
here is incomplete egs list

- found but not PRed task
  - timit/asr1 https://github.com/sknadig/espnet/tree/timit/egs/timit/asr1 @sknadig
  - jsut/asr1 https://github.com/ShigekiKarita/espnet/tree/jsut/egs/just/asr1 @ShigekiKarita 
- missing but important task (what kaldi already has)
  - tedlium v3 (large free dataset)
  - common voice (large free dataset)
  - aurora4 (noisy dataset)
  - libri tts (largest free TTS dataset?) https://www.openslr.org/60/

**Comment by sknadig** [comment]
I'll submit a PR for the TIMIT recipe.
It has scripts for both phoneme and character level decoding.

**Comment by ShigekiKarita** [comment]
@sknadig cool. thanks!

**Comment by Rpersie** [comment]
Anyone has run  the PR #120 "added multichannel end-to-end setup" ? I run the PR #120 but I can not get the reasonable results as the description in the paper.
![image](https://user-images.githubusercontent.com/12673020/56873474-25cddb80-6a65-11e9-8eba-16101aba40d8.png)

**Comment by sw005320** [comment]
@Rpersie, I strongly recommend you to try https://github.com/espnet/espnet/blob/v.0.4.0/egs/chime4/asr1_multich

**Comment by kan-bayashi** [comment]
@Masao-Someki will try to make `tedlium_v3` recipe.

**Comment by ShigekiKarita** [comment]
I met many people  concerning about decoding speed in CPU. How about making that op in native C++/CUDA. I know how to do that in pytorch. Do you have any idea? @Fhrozen @jnishi  https://github.com/tensorflow/lingvo/blob/master/lingvo/core/ops/beam_search_step_op_kernels.cc

**Comment by ShigekiKarita** [comment]
maybe some guys are interested in transformer TTS cc: @kan-bayashi @r9y9

**Comment by jnishi** [comment]
You can use ChainerX from Chainer v 6.0.0. ChainerX is native implementation. Chainer code can be run with ChainerX. https://docs.chainer.org/en/v6.0.0/chainerx/tutorial/index.html#run-your-chainer-code-with-chainerx 
If full code of Chainer backend is replaced to ChainerX, it is stiill faster than Chainer. https://chainer.org/announcement/2018/12/03/chainerx.html

**Comment by ShigekiKarita** [comment]
Thanks. So there will be a way to port op in next chainer v6. That sounds
great.

2019年5月18日(土) 7:33 jnishi <notifications@github.com>:

**Comment by sw005320** [comment]
Can someone try the transformer #690 and specaugment #734 for Librispeech? I think we'll get the SOTA result.

**Comment by kan-bayashi** [comment]
I can do it after adding missing file (#734).

**Comment by kan-bayashi** [comment]
I’m running CSJ & librispeech recipe with transformer + specaug.
Please wait a few days :)

**Comment by sw005320** [comment]
Also, could someone try to combine the transformer and specaug for the WSJ recipe?

**Comment by ShigekiKarita** [comment]
I'm running that.

**Comment by ShigekiKarita** [comment]
@bobchennan Do you really use `W=5` in specaug? It seems to be much shorter than the paper (W=40 or 80) https://arxiv.org/pdf/1904.08779.pdf

**Comment by ShigekiKarita** [comment]
#722 (Support configuration files). As I talked before, it has no break change. cmd line args are still supported. See new [README.md](https://github.com/espnet/espnet/tree/v.0.4.0#changing-the-configuration) for usage.

Maybe you can check this WSJ diff with YAML files and run.sh for the practical example.
https://github.com/espnet/espnet/commit/b9291ae02dd9914b8d04ba768f9f757825a66691#diff-9d8eca030a0ba69ff8e59bbc1559fc50

**Comment by ShigekiKarita** [comment]
I attached "milestone" to the PRs listed in https://github.com/espnet/espnet/issues/650#issue-421526628
You can check opened and closed PR in v.0.4.0 here https://github.com/espnet/espnet/milestone/1
If you find something missing, feel free to attach it to the milestone.

**Comment by sw005320** [comment]
Can someone work on transformer LM? I think it's a good addition.

**Comment by ShigekiKarita** [comment]
Does anyone use pytorch 0.4.1? When we support pytorch 1.1.0, we should drop pytorch 0.4.1.

**Comment by ShigekiKarita** [comment]
Also we should drop Ubuntu 14.04 support because of EOL. We should use Ubuntu 16.04 (xenial).

PS: I found shellcheck in xenial becomes much more strict. We need some help to update run.sh. See #110

**Comment by sw005320** [comment]
> Does anyone use pytorch 0.4.1? When we support pytorch 1.1.0, we should drop pytorch 0.4.1.

We still have some problems on warp CTC in pytorch 1.1.0, which is critical for the WSJ-2mix recipe. I want to keep pytorch 0.4.1 unless we fix the issue.

**Comment by ShigekiKarita** [comment]
I could not find any problem on pytorch 1.1.0 and warp-ctc in my recent CI updates https://github.com/espnet/espnet/pull/763
Is there any reported issue?

**Comment by sw005320** [comment]
https://github.com/espnet/espnet/issues/669

Also, @simpleoier confirmed the difference between these versions.
I also found that the final CTC loss is very different between these versions.
@simpleoier, could you summarize your observations?

**Comment by ShigekiKarita** [comment]
Can you add unit tests that fails in pytorch 1.x?

**Comment by sw005320** [comment]
Sure, but the problem may be more difficult because the initial ranges of loss values are similar but they gradually become different.

**Comment by ShigekiKarita** [comment]
OK... Have you ran training without ctc loss? If they (0.4 and 1.0) still result different then, it might not be caused by ctc.

**Comment by sw005320** [comment]
Also, see https://github.com/espnet/espnet/issues/700

> OK... Have you ran training without ctc loss? If they (0.4 and 1.0) still result different then, it might not be caused by ctc.

That may be true. I'll check it. But the attention loss looks quite normal...

**Comment by sw005320** [comment]
Can someone make a JSUT ASR recipe?

**Comment by kan-bayashi** [comment]
It seems that now chainer trainer support asynchronous snapshot saving.
https://docs.chainer.org/en/stable/reference/generated/chainer.training.extensions.snapshot.html#chainer.training.extensions.snapshot

It is better to use this function to reduce i/o bottleneck in saving model in future.

**Comment by sw005320** [comment]
Thanks. I added it to our action item list.

**Comment by ShigekiKarita** [comment]
@sw005320 should we create new issue for v.0.4.1?

**Comment by sw005320** [comment]
Yeah, let's make it.
